rp2040-hal/src/pwm.rs

```diff
@@ -1,3 +1,312 @@
+                    }
+
+                    #[doc = "Disables output inversion"]
+                    pub fn clr_inv(&self) {
+                        if ($i % 2 == 0) {
+                            self.csr().write(|w| w.a_inv().clear_bit());
+                        } else {
+                            self.csr().write(|w| w.b_inv().clear_bit());
+                        }
+                    }
+
+                    #[doc = "Sets the top register value"]
+                    pub fn set_top(&self, value: u16) {
+                        self.top().write(|w| unsafe { w.$topi().bits(value) });
+                    }
+
+                    #[doc = "Sets the divmode to div. Use this if you aren't reading a PWM input."]
+                    pub fn divmode_div(&self) {
+                        self.csr().write(|w| w.divmode().div());
+                    }
+
+                    #[doc = "Sets the divmode to level."]
+                    pub fn divmode_level(&self) {
+                        self.csr().write(|w| w.divmode().level());
+                    }
+
+                    #[doc = "Sets the divmode to rise."]
+                    pub fn divmode_rise(&self) {
+                        self.csr().write(|w| w.divmode().rise());
+                    }
+
+                    #[doc = "Sets the divmode to fall."]
+                    pub fn divmode_fall(&self) {
+                        self.csr().write(|w| w.divmode().div());
+                    }
+                }
+
+                impl PwmPin for $PXi {
+                    type Duty = u16;
+
+                    fn disable(&mut self) -> () {
+                        self.csr().write(|w| w.en().clear_bit());
+                    }
+
+                    fn enable(&mut self) -> () {
+                        self.csr().write(|w| w.en().set_bit());
+                    }
+
+                    fn get_duty(&self) -> Self::Duty {
+                        if ($i % 2 == 0) {
+                            self.cc().read().a().bits()
+                        } else {
+                            self.cc().read().b().bits()
+                        }
+                    }
+
+                    fn get_max_duty(&self) -> Self::Duty {
+                        self.top().read().$topi().bits()
+                    }
+
+                    fn set_duty(&mut self, duty: Self::Duty) {
+                        if ($i % 2 == 0) {
+                            self.cc().write(|w| unsafe { w.a().bits(duty) });
+                        } else {
+                            self.cc().write(|w| unsafe { w.b().bits(duty) });
+                        }
+                    }
+                }
+            )+
+        }
+    };
+}
+
+// This is stupidly long because rust's concat_idents feature is only available in nightly builds.
+// TODO: See if this can be condensed by creating an array of idents somehow? (no idea if that's possible)
+pwm!(
+    PWM, pwm, IO_BANK0, io_bank0, PADS_BANK0, pads_bank0, "PWM", [
```

Comment:
I'm not sure what you are doing here with 28 PWM objects. The RP2040 only has 8 PWM channels.

With this, you can try and use both `pwm0` and `pwm16` at the same time but that will conflict and try to use the same PWM channel for 2 different pins.